torch/optim/adam.py

```diff
@@ -393,7 +398,19 @@ def _single_tensor_adam(
         step_t += 1
 
         if weight_decay != 0:
```

Comment:
is this causing gpu->cpu copy for a differentiable weight_decay value on gpu device? or is weight_decay always supposed to be a CPU tensor?

Reply:
I designed the differentiable weight decay paths for 0d CPU scalars! For this case, if weight_decay was on a gpu it would still work since the 2nd param in addcmul can either be a tensor (on the same device as grad and param), or a cpu scalar (with grad and param both being on cpu or cuda)

Reply:
Okay! I was just a bit worried that this `if weight_decay` would trigger GPU->CPU copy

Reply:
I don't believe so as I previously changed addcmul to check if tensor2 is a cpuscalar and, it if is, to just to use its value #143264